### ISSUE TITLE
Change of array content not recognized

### ISSUE THREAD

**Issue by svelue** [dotnet/efcore#23313]
I'm using the latest EF Core 5 release and observed a changed behaviour regarding change tracking of array elements.

One of my entites has a DateTime[] property which is converted into a string and stored in the db. While this was working with the latest EF Core 3 release, EF Core 5 does not detect changes in the array. Changing the whole array is detected though. 

I made a small repository with sample applications but the entity and converter code is shown below.

```
    public class Entity
    {
        public Entity()
        {
            Times = new DateTime[3];
        }

        public int Id { get; set; }
        public int Count { get; set; }
        public DateTime[] Times { get; set; }
    }
```

```
        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<Entity>()
                .Property(_ => _.Times)
                .HasConversion(dt => string.Join(";", dt),
                    s => s.Split(";", StringSplitOptions.RemoveEmptyEntries).Select(DateTime.Parse).ToArray());
        }
```

```
        var entity = context.Entities.First();
        entity.Count++;
        entity.Times[1] = DateTime.Now;
        contest.SaveChanges();
```
In the above code modified Count is stored in the db while Times is not.

Full example: [https://github.com/svelue/EfCore5ChangeTrackingBug](https://github.com/svelue/EfCore5ChangeTrackingBug)

### Include provider and version information

EF Core version: 5.0
Database provider: (all tested (InMemory, Npgsql))
Target framework: (.NET 5.0)
Operating system: Windows 10
IDE: (Visual Studio 2019 16.8)

**Comment by ajcvickers** [comment]
@svelue I think the behavior you are seeing with 3.1 is actually a bug. In 3.1, value converters were not used by the in-memory provider. We fixed this for 5.0--see #14042 and related issues. However, in your case, the code for 3.1 is making use of the fact that the actual entity value is being mutated in the in-memory database, rather than the value in the in-memory database being a copy created when SaveChanges is called.

You can use something like:

```C#
Console.WriteLine($"Property modified: {context.Entry(dbEntity).Property(e => e.Times).IsModified}");
```

to see that the property is never marked as modified on either 3.1 or 5.0, meaning that changes are never detected. This is the expected behavior when using a collection type without also using a value comparer. See https://docs.microsoft.com/en-us/ef/core/modeling/value-comparers for more details and example code.

**Comment by svelue** [comment]
@ajcvickers Thx for that clarification! I never stumbled across the fact that list types are compared by reference only, adding a value comparer works like expected.
Since that code was not in production yet the 3.1 bug were not noticed as we test against an in memory db. At least it was a bug, even if it wasn't in the version I thought it is :D
